src/climate_learn/data/climate_dataset/era5_module.py

```diff
@@ -1,94 +1,235 @@
-    args_class: Callable[..., ERA5Args] = ERA5Args
+    _args_class: Callable[..., ERA5Args] = ERA5Args
 
     def __init__(self, data_args: ERA5Args) -> None:
         super().__init__(data_args)
         self.root_dir: str = data_args.root_dir
         self.years: Iterable[int] = data_args.years
 
-    def setup(self) -> None:
-        self.constant_names: Sequence[str] = []
-        self.data_dict: Dict[str, xr.core.dataarray.DataArray] = self.load_from_nc(
-            self.root_dir
+    def get_file_name_from_variable(self, var: str) -> str:
+        if var in SINGLE_LEVEL_VARS or var in PRESSURE_LEVEL_VARS:
+            return var
+        else:
+            return "_".join(var.split("_")[:-1])
+
+    def set_lat_lon(self, year: int) -> None:
+        # lat lon is stored in each of the nc files, just need to load one and extract
+        dir_var = os.path.join(
+            self.root_dir, self.get_file_name_from_variable(self.variables[0])
         )
-        self.set_lat_lon()
+        ps = glob.glob(os.path.join(dir_var, f"*{year}*.nc"))
+        xr_data = xr.open_mfdataset(ps, combine="by_coords")
+        self.lat: numpy.ndarray = xr_data["lat"].values
+        self.lon: numpy.ndarray = xr_data["lon"].values
 
-    def load_from_nc(self, data_dir: str) -> Dict[str, xr.core.dataarray.DataArray]:
-        self.constant_names = [
-            name for name in self.variables if NAME_TO_VAR[name] in CONSTANTS
+    def setup_metadata(self, year: int) -> None:
+        ## Prevent setup if already called before
+        if hasattr(self, "lat") and hasattr(self, "lon"):
+            return
+        self.set_lat_lon(year)
+
+    def setup_constants(self, data_dir: str) -> None:
+        ## Prevent loading constants if already loaded
+        if hasattr(self, "constant_names"):
+            return
+        self.constant_names: Sequence[str] = [
+            name
+            for name in self.variables
+            if NAME_TO_VAR[self.get_file_name_from_variable(name)] in CONSTANTS
         ]
-        self.constants: Dict[str, xr.core.dataarray.DataArray] = {}
+        self.constants: Dict[str, torch.tensor] = {}
         if len(self.constant_names) > 0:
             ps = glob.glob(os.path.join(data_dir, "constants", "*.nc"))
             all_constants = xr.open_mfdataset(ps, combine="by_coords")
             for name in self.constant_names:
-                self.constants[name] = all_constants[NAME_TO_VAR[name]]
+                self.constants[name] = torch.tensor(
+                    all_constants[NAME_TO_VAR[name]].values
+                )
 
-        non_const_names = [
+        # remove constants from variables
+        self.variables = [
             name for name in self.variables if name not in self.constant_names
         ]
-        data_dict = {}
-        for name in non_const_names:
+
+    def initialize_data_dict(
+        self,
+    ) -> Tuple[Dict[str, torch.tensor], Dict[str, Sequence[str]]]:
+        if hasattr(self, "data_dict"):
```

Comment:
The reason why I have to do this in this scenario is because of multi-level variables.

We allow user to write `geopotential` as a variable. In reality we have multiple levels and thus when we populate `data_dict`, its keys are `geopotential_250`, `geopotential_500`, etc and not `geopotential`. This information about new keys should also be passed to the `Task` class so that it knows that `variables` has been updated from `geopotential` to `geopotential_250`, `geopotential_500`.

But once the task has been updated with this new information. We don't really need to do the entire for loop and just instantiate an empty `data_dict` based on the previous keys.

This is not the most clean solution but supporting both `geopotential` and `geopotential_500` is tricky at the same time.

Reply:
What if we do a multilevel dictionary? Like:
```python
{
    "multi-level-variable": {
        level_1: ...,
        level_2: ...,
    },
    "single-level-variable": ...
}
```
I think separating the variable name from the pressure level as distinct keys might be more clear than simply separating them by an underscore in the same key. Is this possible or does it make the code easier to implement?

Reply:
We can do a multi-level dictionary but isn't it specific to `ERA5`? I can imagine in future someone adding a `ClimateDataset` where instead of levels we have something else.

Thus, `Task` should only know `in_vars` and `out_vars` and it becomes task of `ClimateDataset` (in this case `ERA5`) to return `raw_data` in a format that `Task` can operate on.

Reply:
Now whether `in_vars` and `out_vars` should be updated from `geopotential` to `geopotential_250`, ..., etc or not that is a design choice. 

In the former case, `.setup()` logic would involve creating [`variables_to_update`](https://github.com/aditya-grover/climate-learn/blob/ff7f0a6c5d3991a2fd88979ff0afd9a89b7bbd18/src/climate_learn/data/climate_dataset/era5_module.py#L86) and then passing it on to the `Task`'s [`setup()`](https://github.com/aditya-grover/climate-learn/blob/ff7f0a6c5d3991a2fd88979ff0afd9a89b7bbd18/src/climate_learn/data/task/task.py#L23). Now, when calling [`create_inp_out()`](https://github.com/aditya-grover/climate-learn/blob/ff7f0a6c5d3991a2fd88979ff0afd9a89b7bbd18/src/climate_learn/data/task/downscaling.py#L50), it becomes super easy to extract data out of the [`raw_data` ](https://github.com/aditya-grover/climate-learn/blob/ff7f0a6c5d3991a2fd88979ff0afd9a89b7bbd18/src/climate_learn/data/task/downscaling.py#L52) as it's keys are `geopotential_250`, etc.

In the later case, it would become the job of `ClimateDataset`'s [`get_item()`](https://github.com/aditya-grover/climate-learn/blob/ff7f0a6c5d3991a2fd88979ff0afd9a89b7bbd18/src/climate_learn/data/climate_dataset/era5_module.py#L216) to somehow wrap multi-level-variables into one variable. So that `Task` can extract the data out. But again `Task` would have to convert it back to multiple tensors. Unfortunately, in this case it doesn't know how to split it.